markets/storageadapter/client.go

```diff
@@ -264,7 +264,7 @@ func (c *ClientNodeAdapter) OnDealExpiredOrSlashed(ctx context.Context, dealID a
 	// and the chain has advanced to the confidence height
 	stateChanged := func(ts *types.TipSet, ts2 *types.TipSet, states events.StateChange, h abi.ChainEpoch) (more bool, err error) {
 		// Check if the deal has already expired
-		if sd.Proposal.EndEpoch <= ts2.Height() {
+		if ts2 == nil || sd.Proposal.EndEpoch <= ts2.Height() {
```

Comment:
ts2 == nil when there is a timeout which is handled differently on line 273

Reply:
Line 273 is looking at `states` not `ts2`:
`if states == nil { ...`

In reality ts2 should never be nil, it seems like an issue in the events API.
But I think it's better to make this quick fix to make sure there are no further crashes until we track down the events API issue.

Reply:
It's set to nil when the timeout epoch is reached without seeing the tipset you were looking for. Clunky api but at least its documented.

Reply:
It looks like the documentation is out of date 😭  The `oldState` and `newState` parameters have since been changed to a single `states` parameter:
https://github.com/filecoin-project/lotus/blob/c11803ad55a70986c39b0d2132579ecb1a9d5b21/chain/events/events_called.go#L401

I opened an issue: https://github.com/filecoin-project/lotus/issues/5218